crates/virtio-queue-ser/src/state.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+use std::num::Wrapping;
+
+use serde::{Deserialize, Serialize};
+use versionize::{VersionMap, Versionize, VersionizeResult};
+use versionize_derive::Versionize;
+use virtio_queue::QueueState;
+use vm_memory::GuestAddress;
+
+/// /// Wrapper over a `QueueState` that has serialization capabilities.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize, Versionize)]
+pub struct QueueStateSer {
+    /// The maximum size in elements offered by the device.
+    pub max_size: u16,
+    /// Tail position of the available ring.
+    pub next_avail: Wrapping<u16>,
+    /// Head position of the used ring.
+    pub next_used: Wrapping<u16>,
+    /// VIRTIO_F_RING_EVENT_IDX negotiated.
+    pub event_idx_enabled: bool,
+    /// The last used value when using VIRTIO_F_EVENT_IDX.
+    pub signalled_used: Option<Wrapping<u16>>,
+    /// The queue size in elements the driver selected.
+    pub size: u16,
+    /// Indicates if the queue finished with the configuration.
+    pub ready: bool,
+    /// Guest physical address of the descriptor table.
+    pub desc_table: u64,
+    /// Guest physical address of the available ring.
+    pub avail_ring: u64,
+    /// Guest physical address of the used ring.
+    pub used_ring: u64,
+}
+
+// The following `From` implementations can be used to convert from a `QueueStateSer` to the
+// `QueueState` from the base crate and vice versa.
+impl From<&QueueStateSer> for QueueState {
+    fn from(state: &QueueStateSer) -> Self {
+        QueueState {
+            max_size: state.max_size,
+            next_avail: state.next_avail,
+            next_used: state.next_used,
+            event_idx_enabled: state.event_idx_enabled,
+            signalled_used: state.signalled_used,
+            size: state.size,
+            ready: state.ready,
+            desc_table: GuestAddress(state.desc_table),
+            avail_ring: GuestAddress(state.avail_ring),
+            used_ring: GuestAddress(state.used_ring),
+        }
+    }
+}
+
+impl From<&QueueState> for QueueStateSer {
+    fn from(state: &QueueState) -> Self {
```

Comment:
I wonder if there is anything that we need to check here. Is there any invalid state that would otherwise not be able to get into when just initializing a Queue from scratch?